docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+
+## Decision
+
+We will develop a Kubernetes [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) in _Go_ that will install, configure and manage add-on applications in the cluster. The operator will extend the Kubernetes API through Custom Resource Definitions (CRD) and will allow users to declaritively manage add-ons with a Custom Resource (CR). 
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
```

Comment:
One advantage I see with Flux is that it provides a code implementation for reconciling HelmCharts and includes watches and finalizers, as demonstrated in their [helmrelease_controller.go](https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go) file on GitHub. 

However, it comes with the limitation that we cannot modify it according to our specific needs. The controller is generic and may not work effectively with all HelmCharts. By implementing our own solution per Addon, we can automate solutions based on the specific requirements of each Addon. This allows us to address specific needs and requirements more accurately, as a generic implementation will not perform in the same way. It's important to note that a generic automation for a HelmChart will never be able to handle specific issues, workarounds, or tailor solutions that are commonly done in KURL for what we ship.

Reply:
the controller does have a dependsOn property

https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies

i can see us taking advantage of this in conjunction with helm tests for creating pre-requirement checks

Reply:
> It's important to note that a generic automation for a HelmChart will never be able to handle specific issues, workarounds, or tailor solutions that are commonly done in KURL for what we ship

Can you provide an example of what you mean by this? Right now it's just a high level specific > generic and I'm not clear that's true in this case.

Reply:
@chris-sanders,

> 
Can you provide an example of what you mean by this? Right now it's just a high level specific > generic and I'm not clear that's true in this case.

Certainly, I can provide an example. In the following KURL code, we delete resources as a workaround for certain issues: [Link to KURL code](https://github.com/replicatedhq/kURL/blob/b3e3c02ffb04334103c8ad0f80ba6938e2036694/addons/rook/1.9.12/install.sh#L75C4-L81)

However, with Flux, situations like these could trigger the Flux control loop to run again, potentially overwriting our changes since it's designed to ensure the HelmChart is properly installed. Another risk is that if both we and Flux try to update and monitor the same resources, we could encounter race conditions.

I hope that clarifies my concerns.